New Database Trigger Options (#611)
mongodb/docs-app-services: source/triggers/database-triggers.txt

   <serverless-caveats>` or :ref:`{+adf-instance+}
   <data-federation-caveats>` because they do not support change streams.

You control which operations cause a trigger to fire as well as what
happens when it does. For example, you can run a function whenever a
specific field of a document is updated. The function can access the
entire change event, so you always know what changed. You can also pass
the change event to :ref:`AWS EventBridge <aws-eventbridge>` to handle
the event outside of Atlas.

Triggers support :manual:`$match </reference/operator/aggregation/match>` 
expressions to filter change events and :manual:`$project </reference/operator/aggregation/project>` 
expressions to limit the data included in each event.

.. _trigger-recursion:

.. warning::

   It is possible to configure triggers that cause other triggers to fire on the 
   same collection, database, or cluster, resulting in recursion. Examples 
   include a database trigger writing to a collection within the same 
   database, or a cluster-level logger writing database logs to another database 
   in the same cluster.

>    

Reviewer: I think we may want to explicitly call out log forwarders as well here. Since, they are a first class resource that our users may also be using that could cause this issue.

Author: Isn't that just a specific case of a process writing to a monitored database and/or collection?

Reviewer: You are correct, but just wanted to make customer's slightly more conscious if they didn't make that connection.

Reviewer: Talked offline in the project channel with team and Laura, we would like to go with.

```suggestion
.. warning::

   In deployment and database level triggers, it is possible to configure triggers 
   in a way that causes other triggers to fire, thereby resulting in recursion. 
   Examples include a database-level trigger writing to a collection within the 
   same database, or a cluster-level logger or log forwarder writing logs to 
   another database in the same cluster.
   
```

Reviewer: @MongoCaleb Could we get an update on this since it should be going in tomorrow.

~~~~~~~~~~~~~~~~~~~~~~

Within the :guilabel:`Trigger Source Details` section, you first select the 
:guilabel:`Source Type`, based on the level of granularity you want. Your 
options are:

       - :guilabel:`Collection`, when a change occurs on a specified collection
       - :guilabel:`Database`, when a change occurs on any collection in a 
         specified database
       - :guilabel:`Deployment`, when deployment changes occur on a specified 
         cluster. If you select the Deployment source type, the following 
         databases are **not watched** for changes:

         - The admin databases `admin`, `local`, and `config`
         - The sync databases `__realm_sync` and `__realm_sync_<app_id>`
         
Depending on which source type you are using, the additional options differ. The 
following table describes these options.

.. list-table::
   :header-rows: 1
   :widths: 15 30

   * - Source Type
     - Options

   * - :guilabel:`Collection`
     - |
       - :guilabel:`Cluster Name`. The name of the MongoDB cluster that the 
           Trigger is associated with.
      
       - :guilabel:`Database Name`. The MongoDB database that contains the watched

>          collection. Optional if the :guilabel:`Collection Name` is specified.

Reviewer: ```suggestion
         collection. Required if the :guilabel:`Collection Name` is specified.
```

Maybe we should add some copy about how a Deployment trigger is when `database and collection` are not specified. Database trigger is when `collection` is not specified, and Collection Trigger when both are specified.

Author: @nathantfrank does this mean that even if the user selects the Collection or Database source type, it will actually be a Deployment type if they leave both `database` and `collection` blank?

Reviewer: Ahhh Sorry I misunderstood what this section was doing.
My current thought is we should just remove that sentence.
 ```suggestion
         collection.
```

       - :guilabel:`Collection Name`. The MongoDB collection to watch.

>        - :guilabel:`Operation Type`. The :ref:`database operation types 

Reviewer: [Q] should this be `database`?

Author: @nathantfrank Sorry, not sure what you're asking here.

Reviewer: Ahhh, I think I'm confusing myself with existing documentation details. I was thinking this should be like `collection operation types` and link to something different.

  databases are **not watched** for changes:

  - The admin databases ``admin``, ``local``, and ``config``
  - The sync databases ``__realm_sync`` and ``__realm_sync_<app_id>``

Depending on which source type you are using, the additional options differ. The 
following table describes these options.

.. list-table::
   :header-rows: 1
   :widths: 15 30

   * - Source Type
     - Options

   * - :guilabel:`Collection`
     - | 

       - :guilabel:`Cluster Name`. The name of the MongoDB cluster that the 
           Trigger is associated with.
      
       - :guilabel:`Database Name`. The MongoDB database that contains the watched
         collection.

       - :guilabel:`Collection Name`. The MongoDB collection to watch. Optional.
         If you leave this option blank, the Source Type changes to "Database."

       - :guilabel:`Operation Type`. The collection operation types 
         that cause the Trigger to fire. Select the operation types you want the 
         trigger to respond to. Options include:
       
         - Create Document
         - Modify Document
         - Rename Document
         - Shard Document

>          - Drop Document

Reviewer: This should be Insert, Update, Replace, and Delete

Author: Doh!

         the ``fullDocument`` field.
       
         .. note::
       
            Regardless of this setting, **Insert** and **Replace** events always 
            include the ``fullDocument`` field. **Delete** events never include 
            the ``fullDocument`` field.
       
       - :guilabel:`Document Preimage`. Disabled for Collection and Database 
         sources to limit unnecessary watches on the cluster for a new 
         collection being created.

   * - :guilabel:`Database`
     - | 

       - :guilabel:`Cluster Name`. The name of the MongoDB cluster that the 
         Trigger is associated with.
       
       - :guilabel:`Database Name`. The MongoDB database to watch. Optional. 
         If you leave this option blank, the Source Type changes to "Deployment."

       - :guilabel:`Operation Type`. The :ref:`database operation types 
         <database-event-operation-types>` that cause the Trigger to fire. 
         Select the operation types you want the  trigger to respond to. 
         Options include:
       
         - Create Collection
         - Modify Collection
         - Rename Collection
         - Shard Collection

>          - Drop Collection

Reviewer: Missing the 2 new sharding types

  cluster. If you select the Deployment source type, the following 
  databases are **not watched** for changes:

  - The admin databases ``admin``, ``local``, and ``config``
  - The sync databases ``__realm_sync`` and ``__realm_sync_<app_id>``
         
Depending on which source type you are using, the additional options differ. The 
following table describes these options.

.. list-table::
   :header-rows: 1
   :widths: 15 30

   * - Source Type
     - Options

   * - :guilabel:`Collection`
     - | 

       - :guilabel:`Cluster Name`. The name of the MongoDB cluster that the 
           Trigger is associated with.
      
       - :guilabel:`Database Name`. The MongoDB database that contains the watched
         collection.

       - :guilabel:`Collection Name`. The MongoDB collection to watch.

       - :guilabel:`Operation Type`. The collection operation types 
         that cause the Trigger to fire. Select the operation types you want the 
         trigger to respond to. Options include:
       
         - Create Collection
         - Modify Collection
         - Rename Collection
         - Shard Collection

>          - Drop Collection

Reviewer: As mentioned in our chat, these are actually Database+ change event types.

         the latest :manual:`majority-committed </reference/read-concern-majority/>`
         version of the modified document *after* the change was applied in
         the ``fullDocument`` field.
       
         .. note::
       
            Regardless of this setting, **Insert** and **Replace** events always 
            include the ``fullDocument`` field. **Delete** events never include 
            the ``fullDocument`` field.
       
       - :guilabel:`Document Preimage`. Disabled for Collection and Database 
         sources to limit unnecessary watches on the cluster for a new 
         collection being created.

   * - :guilabel:`Database`
     - | 

       - :guilabel:`Cluster Name`. The name of the MongoDB cluster that the 
         Trigger is associated with.
       
       - :guilabel:`Database Name`. The MongoDB database to watch. Optional. 
         If you leave this option blank, the Source Type changes to "Deployment," 
         unless you are on a shared tier, in which case App Services will not 
         let you save the trigger.

       - :guilabel:`Operation Type`. The :ref:`database operation types 

>          <database-event-operation-types>` that cause the Trigger to fire. 

Reviewer: [Q] Only this Database row has a link, do we want to have a link for `Collection` and `Deployment` as well?

.. list-table::
   :header-rows: 1
   :widths: 30 15

   * - Operation Type
     - Description

   * - **Insert Document** (All trigger types)
     - Represents a new document added to the collection.

   * - **Update Document** (All trigger types)
     - Represents a change to an existing document in the collection.

   * - **Delete Document** (All trigger types)
     - Represents a document deleted from the collection.

   * - **Replace Document** (All trigger types)
     - Represents a new document that replaced a document in the collection.

>    * - **Create_Collection** (Database and Deployment trigger types only)

Reviewer: [Q] Any particular reason we are underscoring here?
